### ISSUE TITLE
Issues with Qt 6.7

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#8046]
Qt 6.7 Beta 1 [was released](https://www.qt.io/blog/qt-6.7-beta1-released). Unfortunately, trying to build wheels with the updated binaries:

```
mkdir wheels-67
cd wheels-67
wget https://files.pythonhosted.org/packages/36/91/50982f28730c893a8c68b0e92debc761e8a475cf1df686dee22a4c255316/PyQt6_WebEngine-6.6.0-cp37-abi3-manylinux_2_28_x86_64.whl
wget https://files.pythonhosted.org/packages/0b/5a/51f4762b9f314b5577d17704bc1280532a725ba359d6cc177ab6de692035/PyQt6-6.6.1-cp38-abi3-manylinux_2_28_x86_64.whl
python3 -m venv .venv
.venv/bin/pip install pyqt-builder
.venv/bin/pyqt-bundle --verbose --ignore-missing --qt-dir ~/proj/qt/bin6/6.7.0/gcc_64/ PyQt6-6.6.1-cp38-abi3-manylinux_2_28_x86_64.whl
.venv/bin/pyqt-bundle --verbose --ignore-missing --qt-dir ~/proj/qt/bin6/6.7.0/gcc_64/ PyQt6_WebEngine-6.6.0-cp37-abi3-manylinux_2_28_x86_64.whl
rm PyQt6-6.6.1-cp38-abi3-manylinux_2_28_x86_64.whl PyQt6_WebEngine-6.6.0-cp37-abi3-manylinux_2_28_x86_64.whl
cd ../git
python3 scripts/mkvenv.py --dev --venv-dir .venv-qt67 --pyqt-type wheels --pyqt-wheels-dir ../wheels-67
```

fails with:

```
==================== Running Qt smoke test ====================
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/florian/proj/qutebrowser/git/qutebrowser/qt/widgets.py", line 24, in <module>
    from PyQt6.QtWidgets import *
ImportError: /home/florian/proj/qutebrowser/git/.venv-qt67/lib/python3.11/site-packages/PyQt6/QtGui.abi3.so: undefined symbol: _ZN5QFont11tagToStringEj, version Qt_6
```

which is `QFont::tagToString(unsigned int)`. Related: [Remove obsolete font tag helpers and \preliminary tag from documentation (I7e68298b) · Gerrit Code Review](https://codereview.qt-project.org/c/qt/qtbase/+/510993)

Not sure off-hand why PyQt even includes that:

```cpp
// sip/QtGui/qfont.sip
%If (Qt_6_6_0 -)
    static QByteArray tagToString(quint32 tag);
%End
%If (Qt_6_6_0 -)
    static quint32 stringToTag(const char *tagString /Encoding="None"/);
%End
```

given that it's been marked [as preliminary](https://doc.qt.io/qt-6/qfont.html#tagToString) in Qt 6.6...

**Comment by The-Compiler** [comment]
Using PyQt 6.5 instead:

```
rm PyQt6-6.6.1-6.7.0-cp38-abi3-manylinux_2_28_x86_64.whl
wget https://files.pythonhosted.org/packages/1c/36/e59f8f9f14d977cc03ce340270665b98d4fe302b0186cf428b4a15c1cfb5/PyQt6-6.5.3-cp37-abi3-manylinux_2_28_x86_64.whl
.venv/bin/pyqt-bundle --verbose --ignore-missing --qt-dir ~/proj/qt/bin6/6.7.0/gcc_64/ PyQt6-6.5.3-cp37-abi3-manylinux_2_28_x86_64.whl
rm PyQt6-6.5.3-cp37-abi3-manylinux_2_28_x86_64.whl
```

Makes the smoke test work, but then importing QtQml/QtNetwork fails...:

```
>>> from PyQt6 import QtQml
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: /home/florian/proj/qutebrowser/git/.venv-qt67/lib/python3.11/site-packages/PyQt6/QtNetwork.abi3.so: undefined symbol: _ZN16QNetworkDatagram7destroyEP23QNetworkDatagramPrivate, version Qt_6_PRIVATE_API
```

Which is due to `QNetworkDatagram::destroy(QNetworkDatagramPrivate*)`.

Apparently fixed in Qt 6.6.1: [[QTBUG-118229] QNetworkDatagram destructor uses private API - Qt Bug Tracker](https://bugreports.qt.io/browse/QTBUG-118229)

So I suppose for now we're out of luck without building things from source. Enough spelunking for today!

**Comment by ionenwks** [comment]
fwiw qutebrowser *looked* normal when I tried 6.7, not that I used it that much with it nor ran tests

Mentioned that before on IRC, but if end up wanting to build PyQt6 from source, I wrote a basic patch so it can build against 6.7 which worked back then (not that I tested in a while so there could be new problems, will be waiting for 6.7.0-rc to really review):
https://github.com/gentoo/gentoo/blob/8d2240e/dev-python/PyQt6/files/PyQt6-6.6.1-qt670.patch

**Comment by The-Compiler** [comment]
With the [current snapshot](https://www.riverbankcomputing.com/static/Downloads/PyQt6/ChangeLog-6.7.0.dev2403171504), I tried:

```
yay -S icu73   # needed by PyQt binaries, took around 3.5mins on my machine
mkdir wheels-67
cd wheels-67
wget 'https://www.riverbankcomputing.com/pypi/packages/PyQt6/PyQt6-6.7.0.dev2403171504-cp38-abi3-manylinux_2_28_x86_64.whl#sha256=5b812a4389dfcc8062d0f22080a7b288f545764e9c8cbafe703e8fe778531071'
wget 'https://www.riverbankcomputing.com/pypi/packages/PyQt6-WebEngine/PyQt6_WebEngine-6.7.0.dev2403171553-cp38-abi3-manylinux_2_28_x86_64.whl#sha256=c362e83447061efd0eb335aa954b42ae5b9267d733b6e500e85aaabf51ed700b'
python3 -m venv .venv
.venv/bin/pip install pyqt-builder
.venv/bin/pyqt-bundle --verbose --ignore-missing --qt-dir ~/proj/qt/bin6/6.7.0/gcc_64/ PyQt6-6.7.0.dev2403171504-cp38-abi3-manylinux_2_28_x86_64.whl
.venv/bin/pyqt-bundle --verbose --ignore-missing --qt-dir ~/proj/qt/bin6/6.7.0/gcc_64/ PyQt6_WebEngine-6.7.0.dev2403171553-cp38-abi3-manylinux_2_28_x86_64.whl
rm PyQt6-6.7.0.dev2403171504-cp38-abi3-manylinux_2_28_x86_64.whl PyQt6_WebEngine-6.7.0.dev2403171553-cp38-abi3-manylinux_2_28_x86_64.whl
cd ../git
python3 scripts/mkvenv.py --dev --venv-dir .venv-qt67 --pyqt-type wheels --pyqt-wheels-dir ../wheels-67
```

which still fails with:

```pytb
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/florian/proj/qutebrowser/git/qutebrowser/qt/widgets.py", line 24, in <module>
    from PyQt6.QtWidgets import *
ImportError: /home/florian/proj/qutebrowser/git/.venv-qt67/lib/python3.11/site-packages/PyQt6/QtGui.abi3.so: undefined symbol: _ZN5QFont11tagToStringEj, version Qt_6
```

so I guess I should just build PyQt from source.

**Comment by toofar** [comment]
I raised a bug here about the vulkan backend breaking that @ionenwks reported on IRC: [[QTBUG-123607] Vulkan backend rendering only black on X11 - Qt Bug Tracker](https://bugreports.qt.io/browse/QTBUG-123607)

**Comment by The-Compiler** [comment]
Finally running tests against a (debug) build from source now. Not looking too bad. As usual, issues I encounter below with one comment per issue.

**Comment by The-Compiler** [comment]
# Chromium debug check failures

```diff
diff --git i/chromium/components/viz/service/frame_sinks/frame_sink_manager_impl.cc w/chromium/components/viz/service/frame_sinks/frame_sink_manager_impl.cc
index 260b600f93b..ab411992085 100644
--- i/chromium/components/viz/service/frame_sinks/frame_sink_manager_impl.cc
+++ w/chromium/components/viz/service/frame_sinks/frame_sink_manager_impl.cc
@@ -96,10 +96,10 @@ FrameSinkManagerImpl::~FrameSinkManagerImpl() {
 
   // All mojom::CompositorFrameSinks and BeginFrameSources should be deleted by
   // this point.
-  DCHECK(sink_map_.empty());
-  DCHECK(root_sink_map_.empty());
-  DCHECK(cached_back_buffers_.empty());
-  DCHECK(registered_sources_.empty());
+  // DCHECK(sink_map_.empty());
+  // DCHECK(root_sink_map_.empty());
+  // DCHECK(cached_back_buffers_.empty());
+  // DCHECK(registered_sources_.empty());
 
   surface_manager_.RemoveObserver(this);
   surface_manager_.RemoveObserver(&hit_test_manager_);
```

not sure what's going on there, but probably not worth investigating further.

**Comment by The-Compiler** [comment]
# New error messages

```diff
diff --git a/pytest.ini b/pytest.ini
index 5c4c2095a..883e6d936 100644
--- a/pytest.ini
+++ b/pytest.ini
@@ -64,6 +64,8 @@ qt_log_ignore =
     # Seen in Qt 6.6.2 on CI, https://github.com/qutebrowser/qutebrowser/pull/8106#issuecomment-1952320663
     ^QDBusConnection: couldn't handle call to Notify, no slot matched
     ^QDBusConnection: couldn't handle call to CloseNotification, no slot matched
+    # Qt 6.7
+    ^Path override failed for key base::DIR_APP_DICTIONARIES and path '.*/qtwebengine_dictionaries'
 xfail_strict = true
 filterwarnings =
     error
diff --git a/tests/end2end/fixtures/quteprocess.py b/tests/end2end/fixtures/quteprocess.py
index 98f7eff4d..353c0d4a0 100644
--- a/tests/end2end/fixtures/quteprocess.py
+++ b/tests/end2end/fixtures/quteprocess.py
@@ -213,6 +213,11 @@ def is_ignored_chromium_message(line):
         # [9895:9983:0904/043039.500565:ERROR:gpu_memory_buffer_support_x11.cc(49)]
         # dri3 extension not supported.
         "dri3 extension not supported.",
+
+        # Qt 6.7 debug build
+        # [44513:44717:0325/173456.146759:WARNING:render_message_filter.cc(144)]
+        # Could not find tid
+        "Could not find tid",
     ]
     return any(testutils.pattern_match(pattern=pattern, value=message)
                for pattern in ignored_messages)

```

The QtWebEngine-dictionaries one maybe warrants some further investigation? Though I've heard people say that spell checking works just fine.

**Comment by The-Compiler** [comment]
# Accessing statics from tests

Accessing certain statics at import time now seems to fail without a QApplication.

Context: [Q_APPLICATION_STATIC: Assert existence of a QCoreApplication instance (Id9d4a346) · Gerrit Code Review](https://codereview.qt-project.org/c/qt/qtbase/+/495239)

Note the:

>  Without the assertion only a message about invalid nullptr used with QObject::connect() is emitted. This message is much less informative and does not cause tests to fail.

Which I think we've seen with debug builds before? So looks like one problem gone now possibly!

```diff
commit b025775fdeb02ecf6903e6e3721744355ff975f3
Author: Florian Bruhin <me@the-compiler.org>
Date:   Mon Mar 25 17:25:32 2024 +0100

    tests: Avoid accessing Qt statics at import time
    
    With a Qt 6.7 developer build, the tests fail with:
    
        ASSERT failure in QtGlobalStatic::ApplicationHolder<QAS>::PlainType*
        QtGlobalStatic::ApplicationHolder<QAS>::pointer() [with QAS =
        {anonymous}::Q_QAS_qtlsbLoader; PlainType = QFactoryLoader]: "The
        application static was used without a QCoreApplication instance", file
        .../qtbase/src/corelib/kernel/qapplicationstatic.h, line 54
    
        Fatal Python error: Aborted
    
        [...]
    
        Current thread 0x00007c18bb3f3740 (most recent call first):
        File ".../tests/unit/browser/webkit/test_certificateerror.py", line 23 in <module>

diff --git a/tests/unit/browser/webkit/test_certificateerror.py b/tests/unit/browser/webkit/test_certificateerror.py
index 2feb7dcf1..7f5cce9c6 100644
--- a/tests/unit/browser/webkit/test_certificateerror.py
+++ b/tests/unit/browser/webkit/test_certificateerror.py
@@ -18,15 +18,15 @@ def errorString(self):
         return self.msg
 
 
-@pytest.mark.parametrize('errors, expected', [
+@pytest.mark.parametrize('error_factories, expected', [
     (
-        [QSslError(QSslError.SslError.UnableToGetIssuerCertificate)],
+        [lambda: QSslError(QSslError.SslError.UnableToGetIssuerCertificate)],
         ['<p>The issuer certificate could not be found</p>'],
     ),
     (
         [
-            QSslError(QSslError.SslError.UnableToGetIssuerCertificate),
-            QSslError(QSslError.SslError.UnableToDecryptCertificateSignature),
+            lambda: QSslError(QSslError.SslError.UnableToGetIssuerCertificate),
+            lambda: QSslError(QSslError.SslError.UnableToDecryptCertificateSignature),
         ],
         [
             '<ul>',
@@ -37,13 +37,13 @@ def errorString(self):
     ),
 
     (
-        [FakeError('Escaping test: <>')],
+        [lambda: FakeError('Escaping test: <>')],
         ['<p>Escaping test: &lt;&gt;</p>'],
     ),
     (
         [
-            FakeError('Escaping test 1: <>'),
-            FakeError('Escaping test 2: <>'),
+            lambda: FakeError('Escaping test 1: <>'),
+            lambda: FakeError('Escaping test 2: <>'),
         ],
         [
             '<ul>',
@@ -53,8 +53,9 @@ def errorString(self):
         ],
     ),
 ])
-def test_html(stubs, errors, expected):
+def test_html(stubs, error_factories, expected):
     reply = stubs.FakeNetworkReply(url=QUrl("https://example.com"))
+    errors = [factory() for factory in error_factories]
     wrapper = certificateerror.CertificateErrorWrapper(reply=reply, errors=errors)
     lines = [line.strip() for line in wrapper.html().splitlines() if line.strip()]
     assert lines == expected
diff --git a/tests/unit/config/test_configtypes.py b/tests/unit/config/test_configtypes.py
index d908e3ac6..bcd257ed7 100644
--- a/tests/unit/config/test_configtypes.py
+++ b/tests/unit/config/test_configtypes.py
@@ -1874,24 +1874,24 @@ class TestProxy:
     def klass(self):
         return configtypes.Proxy
 
-    @pytest.mark.parametrize('val, expected', [
-        ('system', configtypes.SYSTEM_PROXY),
-        ('none', QNetworkProxy(QNetworkProxy.ProxyType.NoProxy)),
+    @pytest.mark.parametrize('val, expected_factory', [
+        ('system', lambda: configtypes.SYSTEM_PROXY),
+        ('none', lambda: QNetworkProxy(QNetworkProxy.ProxyType.NoProxy)),
         ('socks://example.com/',
-         QNetworkProxy(QNetworkProxy.ProxyType.Socks5Proxy, 'example.com')),
+         lambda: QNetworkProxy(QNetworkProxy.ProxyType.Socks5Proxy, 'example.com')),
         ('socks5://foo:bar@example.com:2323',
-         QNetworkProxy(QNetworkProxy.ProxyType.Socks5Proxy, 'example.com', 2323,
-                       'foo', 'bar')),
+         lambda: QNetworkProxy(
+             QNetworkProxy.ProxyType.Socks5Proxy, 'example.com', 2323, 'foo', 'bar')),
         ('pac+http://example.com/proxy.pac',
-         pac.PACFetcher(QUrl('pac+http://example.com/proxy.pac'))),
+         lambda: pac.PACFetcher(QUrl('pac+http://example.com/proxy.pac'))),
         ('pac+file:///tmp/proxy.pac',
-         pac.PACFetcher(QUrl('pac+file:///tmp/proxy.pac'))),
+         lambda: pac.PACFetcher(QUrl('pac+file:///tmp/proxy.pac'))),
     ])
-    def test_to_py_valid(self, klass, val, expected):
+    def test_to_py_valid(self, klass, val, expected_factory):
         actual = klass().to_py(val)
         if isinstance(actual, QNetworkProxy):
             actual = QNetworkProxy(actual)
-        assert actual == expected
+        assert actual == expected_factory()
 
     @pytest.mark.parametrize('val', [
         'blah',

```

**Comment by The-Compiler** [comment]
# New `QDatastream.Status` entry

```
tests/unit/utils/test_qtutils.py .......................................................F

_______________________________________________________________________________________________________ test_qdatastream_status_count ________________________________________________________________________________________________________

    def test_qdatastream_status_count():
        """Make sure no new members are added to QDataStream.Status."""
        status_vals = testutils.enum_members(QDataStream, QDataStream.Status)
>       assert len(status_vals) == 4
E       AssertionError: assert 5 == 4
E        +  where 5 = len({'Ok': <Status.Ok: 0>, 'ReadCorruptData': <Status.ReadCorruptData: 2>, 'ReadPastEnd': <Status.ReadPastEnd: 1>, 'SizeLimitExceeded': <Status.SizeLimitExceeded: 4>, ...})

tests/unit/utils/test_qtutils.py:232: AssertionError
```

[QDataStream Class | Qt Core 6.7.1](https://doc-snapshots.qt.io/qt6-6.7/qdatastream.html#Status-enum) says:

> `QDataStream::SizeLimitExceeded` (since Qt 6.7): The data stream cannot read or write the data because its size is larger than supported by the current platform. This can happen, for example, when trying to read more that 2 GiB of data on a 32-bit platform.

Test was added in 1e982a9a84784b870a00fba13c689e146d1ffbdc. Looks like `qtutils.check_qdatastream` and the `test_check_qdatastream` / `test_qdatastream_status_count` tests need adjustments.

**Comment by starmonger999** [comment]
Will this take a while? qutebrowser is still broken for those using qt6.7

**Comment by The-Compiler** [comment]
@starmonger999 "this"? "still broken"? None of the above issues are user-facing, and all of them have already been addressed.

**Comment by alzheimerspenguin** [comment]
mine is broken :)

➜  ~ qutebrowser 
Fatal error: PyQt6.QtWidgets is required to run qutebrowser but could not be imported! Maybe it's not installed?

The error encountered was:
/usr/lib/python3.11/site-packages/PyQt6/QtGui.abi3.so: undefined symbol: _ZN5QFont11tagToStringEj, version Qt_6

Please search for the python3 version of PyQt6.QtWidgets in your distributions packages, or see https://github.com/qutebrowser/qutebrowser/blob/main/doc/install.asciidoc

If you installed a qutebrowser package for your distribution, please report this as a bug.

Qt wrapper info:
  PyQt6: success
  PyQt5: not imported
  -> selected: PyQt6 (via autoselect)
➜  ~

**Comment by ionenwks** [comment]
> mine is broken :)
> 
> ➜ ~ qutebrowser Fatal error: PyQt6.QtWidgets is required to run qutebrowser but could not be imported! Maybe it's not installed?
> 
> The error encountered was: /usr/lib/python3.11/site-packages/PyQt6/QtGui.abi3.so: undefined symbol: _ZN5QFont11tagToStringEj, version Qt_6

That's a PyQt6 issue rather than qutebrowser. See earlier posts which talked bout this, aka PyQt6-6.6.1 is not compatible with Qt6.7.0 (or at least not without patching+rebuilding like I did in [an earlier comment](https://github.com/qutebrowser/qutebrowser/issues/8046#issuecomment-1880155731)). Haven't tried it but more recently there's a [dev snapshot](https://www.riverbankcomputing.com/pypi/simple/pyqt6/) for PyQt6-6.7.0 too.

**Comment by alzheimerspenguin** [comment]
Right right yes I understand; I posted vaguely in response to "not user facing."
I'm a user and my browser doesn't work.

Not a demand just letting you know.

**Comment by The-Compiler** [comment]
I was referring to the qutebrowser specific issues in the comments above. What you're seeing is a problem with your setup and not qutebrowser. Your browser also wouldn't work if your Linux wouldn't boot, and then here wouldn't be the correct place to report that either.

**Comment by iambumblehead** [comment]
tldr for anyone using Alpine or postmarketOS; qutebrowser does not start (and shows [the error above][2]) because the [py3-qt6-6.6.1-r1][0] must be updated to qt-6.7. 

[0]: https://pkgs.alpinelinux.org/package/edge/community/armhf/py3-qt6
[1]: https://pkgs.alpinelinux.org/package/edge/community/aarch64/py3-pyqt6-webengine
[2]: https://github.com/qutebrowser/qutebrowser/issues/8046#issuecomment-2053569668
